Add a new [build_options] category (#13)
LedgerHQ/ledgered: doc/utils/manifest.md

> [tests]
> unit_directory = "./unit-tests/"
> pytest_directory = "./tests/"
> 
> [use_cases]
> debug = "DEBUG=1"
> test = "DEBUG=1"

Reviewer: ```suggestion
[tests]
unit_directory = "./unit-tests/"
pytest_directory = "./tests/"

[tests.dependencies.testing_develop]
[
{ url = "https://github.com/<owner>/<app-repository>", ref = "develop", use_case = "debug" },
{ url = "https://github.com/<owner>/<other-app-repository>", ref = "develop" },
]

[use_cases]
debug = "DEBUG=1"
test = "DEBUG=1"
```

Author: Done

```

> ### Sections

Reviewer 2: Could we clearly separate `Required sections` and `Optional sections`  each into its own markdown section ? ^^ @lpascal-ledger what do you think ?

Author: I don't think there should be any `Required sections`. Let's just use and suggest a common naming and not add more mandatory / enforced stuff in the manifest

| `devices`         | The list of devices on which the application can be built.                                             |

#### `[use_cases]`

This section is optional. It contains metadata helping select build options depending on use cases
The VSCode extension leverages this section to provide alternative build targets.

| Field name   | Description                                                             |
|--------------|-------------------------------------------------------------------------|
| `<use_case>` | Options string: <ul><li>Environment variables definitions for C applications (e.g. `DEBUG=1`)</li><li>Valid Cargo build options for Rust applications (e.g. `--outdir mydir`)</li></ul> |

This specifies that in order to build for `<use_case>`, the options string must be provided in the build command line.
You are free to add any use case you wish to have a VSCode build target for.
> [!WARNING]
> The use case `"default"` refers to a standard build with no option. It is implicit and can't be redefined.

Example:
```
[use_cases] # Coherent build options that make sense for your application
debug = "DEBUG=1"
test = "TESTING=1 TEST_PUBLIC_KEY=1"

> my_variant = "COIN=MY_VARIANT"

Reviewer 3: I'm wondering if variant shouldn't be handled differently?
Cause we might need it for other tools, like CI tests.
Also it will probably need a lot of use_case combination (like if I want my variant + debug ?) How is it going to look in vsCode? e.g. can i choose multiple use_case at the same time?

Author: I suggest seeing if this proposal fits the need, then if needed handle the variant in step 2
Most apps have few variants (less than 3), in this case it looks like:
```
foo = ...
foo_debug = ...
bar = ...
bar_debug = ...
baz = ...
baz_debug = ...
```
6 lines is very acceptable

Extracting the variant from the makefile to the manifest is easy, but then having the makefile read the manifest for the info in a clear way is difficult

Reviewer 3: I was more suggesting having the variant list in both the makefile and the manifest, rather than removing it from the makefile which indeed would be tricky.

Reviewer 3: I definitively think that the variant section would help improve our CI.

Reviewer 3: Note that variants might not be available on all devices...

Reviewer 3: Maybe we can define a variant section with shared variant and device specific variant?

| `unit_directory`   | Path of the directory where unit tests can be found                                            |
| `pytest_directory` | Path of the directory where functional, Python test can be found (`conftest.py` file expected) |

#### `[build_options]`

This section is optional. It contains metadata helping select build options depending on use cases

| Field name   | Description |
|--------------|-------------|

> | `<use_case>` | `OPTIONS`   |

Reviewer 2: Maybe explain a bit more what should be "OPTIONS" (and where it will be used). I guess it will also depend on whether it is a rust or C app. 

For a C app it will be (I guess) env variables passed before calling make : `<OPTIONS> && make` but for Rust apps it will be passed as a cargo option `cargo ledger build nanos -- <OPTIONS>`

| `unit_directory`   | Path of the directory where unit tests can be found                                            |
| `pytest_directory` | Path of the directory where functional, Python test can be found (`conftest.py` file expected) |

#### `[build_options]`

This section is optional. It contains metadata helping select build options depending on use cases

| Field name   | Description                                                             |
|--------------|-------------------------------------------------------------------------|

> | `<use_case>` | `OPTIONS` (ENV variables for C apps, cargo build options for Rust apps) |

Reviewer 2: ```suggestion
| `<use_case>` | Options string : <ul><li>Environment variables definitions for C apps (i.e. `DEBUG=1`)</li><li>Valid Cargo build options for Rust apps (i.e. `--outdir mydir`)</li></ul> |
```

> This specify that in order to build for `<use_case>`, the `OPTIONS` must be provided in the build command line.

Reviewer 2: ```suggestion
This specify that in order to build for `<use_case>`, the options string must be provided in the build command line.
```

| `unit_directory`   | Path of the directory where unit tests can be found                                            |
| `pytest_directory` | Path of the directory where functional, Python test can be found (`conftest.py` file expected) |

#### `[use_cases]`

This section is optional. It contains metadata helping select build options depending on use cases
The VSCode extension leverages this section to provide alternative build targets.

| Field name   | Description                                                             |
|--------------|-------------------------------------------------------------------------|

> | `<use_case>` | Options string : <ul><li>Environment variables definitions for C applications (e.g. `DEBUG=1`)</li><li>Valid Cargo build options for Rust applications (e.g. `--outdir mydir`)</li></ul> |

Reviewer: ```suggestion
| `<use_case>` | Options string: <ul><li>Environment variables definitions for C applications (e.g. `DEBUG=1`)</li><li>Valid Cargo build options for Rust applications (e.g. `--outdir mydir`)</li></ul> |
```

> This specify that in order to build for `<use_case>`, the options string must be provided in the build command line.

Reviewer: ```suggestion
This specifies that in order to build for `<use_case>`, the options string must be provided in the build command line.
```

Author: Done

You are free to add any use case you wish to have a VSCode build target for.

> The use case `default` refers to a standard build with no option. It is implicit and can't be redefined.

Reviewer: ```suggestion
> [!WARNING]
> The use case `"default"` refers to a standard build with no option. It is implicit and can't be redefined.
```

Author: Done

Example:
```

> [use_cases] # Coherent build options that make sense for your application
> debug = "DEBUG=1"
> test = "TESTING=1 TEST_PUBLIC_KEY=1"

Reviewer: ```suggestion
[use_cases] # Coherent build options that make sense for your application
debug = "DEBUG=1"
test = "TESTING=1 TEST_PUBLIC_KEY=1"
variant = "COIN=VARIANT"
```

```

#### `[test.dependencies.<test_use_case>]`

The test.dependencies.* sections are optional. They contain metadata helping building side applications needed for your tests.
You can define as many as you need.
The VSCode extension leverages this field to build test dependencies for Speculos tests or device tests.

| Field name | Description                                                                                       |
|------------|---------------------------------------------------------------------------------------------------|
| `[{<url>, <ref>, <use_case>}]` | A table of dependent applications to build for the `<test_use_case>` use case |

> This specify that in order to prepare for the `test_use_case`, the application located at `<url>` on ref `<ref>` should be compiled with its `<use_case>`.

Reviewer: Inconsistent with the `Field name` header and could be misinterpreted as it is not clear `{<url>, <ref>, <use_case>}` refers to `key=value`.
I don't think we can do without explaining 1. each `tests.dependencies.*` is a list and 2. here are the fields of each element of this list.

~~~suggestion
The syntax to describe a single side application is the following:

| Field name | Description                                                           |
|------------|-----------------------------------------------------------------------|
| `url`      | The URL of the application git repository  |
| `ref`      | The reference to checked out in the repository |
| `use_case` | The `use_case` to use to build the application |

This is to be included in a list of applications attached to a particular testing scenario, for instance:
```toml
[[tests.dependencies.test_use_case]]
url = <url>
ref = <ref>
use_case = <use_case>
```
This specifies that the `test_use_case` scenario needs the application located at `<url>` on ref `<ref>` and should be built according to its `<use_case>`.

This following syntax can be used, as it is lighter and equivalent to the previous one:
```toml
[tests.dependencies]
test_use_case = {url = <url>, ref = <ref>, use_case = <use_case>}
```
~~~

> Do note that the `<use_case>` is the one of the application refered to, not this application.

Reviewer: ```suggestion
> [!INFO]
> Do note that the `<use_case>` is the one of the application referred to, not to this manifest's application. Hence it must be defined on the other application own manifest (or be `"default"`)

> [!INFO]
> The `<use_case>` is optional, and will fallback to `use_case = "default"` if non present.

```

The syntax to describe a single side application is the following:

| Field name | Description                                                           |
|------------|-----------------------------------------------------------------------|
| `url`      | The URL of the application git repository  |
| `ref`      | The reference to checkout in the repository |
| `use_case` | The `use_case` to use to build the application |

This is to be included in a list of applications attached to a particular testing scenario, for instance:
```toml
[[tests.dependencies.test_use_case]]
url = <url>
ref = <ref>
use_case = <use_case>
```
This specifies that the `test_use_case` scenario needs the application located at `<url>` on ref `<ref>` and should be built according to its `<use_case>`.

This following syntax can be used, as it is lighter and equivalent to the previous one:
```toml
[tests.dependencies]
test_use_case = [{url = <url>, ref = <ref>, use_case = <use_case>}]
```

> [!INFO]
> Do note that the `<use_case>` is the one of the application referred to, not to this manifest's application. Hence it must be defined on the other application own manifest (or be `"default"`)

> > [!INFO]

Reviewer 2: This `!INFO` does not seem to be properly displayed when rendering the file on github

Reviewer 2: ![image](https://github.com/LedgerHQ/ledgered/assets/111449109/38adbfca-104c-4672-ad8d-3989252cd09e)

Reviewer 2: Cf. https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

> The `<use_case>` is optional, and will fallback to `use_case = "default"` if non present.

Example for an Ethereum plugin that needs the Ethereum application sideloaded on the device:
```toml
[tests.dependencies] # Set of applications to load on the device (or Speculos) for a given test suite

testing_with_prod = [
  {url = "https://github.com/LedgerHQ/app-ethereum", ref = "master", use_case = "default"},
]

testing_with_latest = [

>   {url = "https://github.com/LedgerHQ/app-ethereum", ref = "develop", use_case = "default"},

Reviewer 3: Maybe we recommend testing with debug? So that by default user that do this will have logs in ragger output?

Author: Changed to use `cal_key_debug`
